### PR TITLE
Feature/bean post processor

### PR DESCRIPTION
**Pull request by ajax-makarytskyi-o**
Bean post processor with annotation were added.

### CODE REVIEW

File: src/main/kotlin/com/makarytskyi/rentcar/service/impl/CarServiceImpl.kt
Comment (ajax-torbin-v) [inline src/main/kotlin/com/makarytskyi/rentcar/service/impl/CarServiceImpl.kt:1]: Pls use same style for methods in one file

File: src/main/kotlin/com/makarytskyi/rentcar/service/impl/CarServiceImpl.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    override fun update(id: String, carRequest: UpdateCarRequest): CarResponse =
+        repository.update(id, UpdateCarRequest.toEntity(carRequest))?.let { CarResponse.from(it) }
+            ?: throw ResourceNotFoundException("Car with id $id is not found")
+
+    override fun getByPlate(plate: String): CarResponse = repository.findByPlate(plate)?.let { CarResponse.from(it) }
+        ?: throw ResourceNotFoundException("Car with plate $plate is not found")
+
+    private fun validatePlate(plate: String) {
+        require(repository.findByPlate(plate) == null) { "Car with plate $plate is already exist" }
+    }
```
Comment (ajax-torbin-v) [inline src/main/kotlin/com/makarytskyi/rentcar/service/impl/CarServiceImpl.kt]: Look at comments for file

File: src/test/kotlin/beanpostprocessors/BeanPostProcessorTests.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        //WHEN
+        val proxyService = postProcessor.postProcessAfterInitialization(service, name)
+
+        //THEN
+        assertNotNull(proxyService)
+        assertNotEquals(proxyService!!::class.java, service::class.java)
+    }
+
+    @Test
+    fun `proxied controller should write in logger`() {
```
Comment (ajax-malov-n) [inline src/test/kotlin/beanpostprocessors/BeanPostProcessorTests.kt]: Update method name

File: src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any {
+        val beanClass = bean::class
+        if (beanClass.findAnnotation<InvocationTracker>() != null) {
+            classes[beanName] = bean::class
+        }
+        return bean
+    }
+
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
```
Comment (ajax-torbin-v) [inline src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt]: Can return object be null here?

File: src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        val originalClassName = classes[beanName]?.simpleName
```
Comment (ajax-torbin-v) [inline src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt]: Naming is kinda confusing
```suggestion
        val simpleClassName = classes[beanName]?.simpleName
```

File: src/test/kotlin/beanpostprocessors/BeanPostProcessorTests.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        //GIVEN
+        val name = "service"
+        postProcessor.postProcessBeforeInitialization(service, name)
+
+        //WHEN
+        val proxyService = postProcessor.postProcessAfterInitialization(service, name)
+
+        //THEN
+        assertNotNull(proxyService)
+        assertNotEquals(proxyService!!::class.java, service::class.java)
```
Comment (ajax-torbin-v) [inline src/test/kotlin/beanpostprocessors/BeanPostProcessorTests.kt]: Refer to comments in bpp to remove !!

File: src/test/kotlin/beanpostprocessors/BeanPostProcessorTests.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        val id = "unexistingId"
+        val name = "service"
+
+        //WHEN
+        postProcessor.postProcessBeforeInitialization(service, name)
+        val proxyService = postProcessor.postProcessAfterInitialization(service, name) as CarService
+
+        //THEN
+        assertThrows<ResourceNotFoundException> { proxyService.getById(id) }
+        assert(proxyService::class != service::class) { "Proxy should be created for the bean" }
```
Comment (ajax-malov-n) [inline src/test/kotlin/beanpostprocessors/BeanPostProcessorTests.kt]: Remove this line

Comment (ajax-torbin-v) [inline src/test/kotlin/beanpostprocessors/BeanPostProcessorTests.kt]: You should use AOP utils to determine, if proxyService is proxy

File: src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                beanClass.java.classLoader,
+                beanClass.java.interfaces,
+                InvocationTrackerHandler(bean, originalClassName)
+            )
+        } ?: bean
+        return res
+    }
+
+
+    class InvocationTrackerHandler(private val bean: Any, private val originalClassName: String?) : InvocationHandler {
```
Comment (ajax-malov-n) [inline src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt]: make this private

File: src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        override fun invoke(proxy: Any?, method: Method?, args: Array<out Any>?): Any? {
+            val startTime = System.currentTimeMillis()
+            return try {
+                val result = method?.invoke(bean, *(args ?: emptyArray()))
+                val endTime = System.currentTimeMillis()
+                log.atInfo()
+                    .setMessage("Method '{}' of class '{}' with arguments: {} returned {} executed in {} ms")
+                    .addArgument(method?.name)
+                    .addArgument(originalClassName)
+                    .addArgument { args?.joinToString(", ", "[ ", " ]") }
```
Comment (ajax-malov-n) [inline src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt]: Prefer passing lamba as a last argument

File: src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                    .addArgument(result)
+                    .addArgument(endTime - startTime)
+                    .addArgument { args?.joinToString(", ", "[ ", " ]") }
+                    .log()
+                result
+
+            } catch (e: InvocationTargetException) {
+                val endTime = System.currentTimeMillis()
+                log.atError()
+                    .setMessage("Method '{}' of class '{}' with arguments: {} finished in {} ms and threw {}: '{}'")
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt]: log.error can accept exception as the last parameter and will log it with stacktrace

**Review by ajax-samondros-b — APPROVED** [review]

File: src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt [hunk 10]
```diff
@@ -57,14 +57,14 @@ internal class InvocationTrackerBeanPostProcessor : BeanPostProcessor {
 
             } catch (e: InvocationTargetException) {
                 val endTime = System.currentTimeMillis()
+
                 log.atError()
-                    .setMessage("Method '{}' of class '{}' with arguments: {} finished in {} ms and threw {}: '{}'")
+                    .setMessage("Method '{}' of class '{}' with arguments: {} finished in {} ms and threw")
```
Comment (ajax-shulepa-a) [inline src/main/kotlin/com/makarytskyi/rentcar/bpp/InvocationTrackerBeanPostProcessor.kt]:
```suggestion
                    .setMessage("Method '{}' of class '{}' with arguments: {} finished with exception in {} ms")
```

**Review by ajax-shulepa-a — APPROVED** [review]

**Pull request merged**